toolchain/check/handle_impl.cpp

```diff
@@ -313,17 +326,22 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
       {.self_id = self_inst_id, .constraint_id = constraint_inst_id}};
 
   // Add the impl declaration.
+  bool valid_redeclaration = true;
   auto lookup_bucket_ref = context.impls().GetOrAddLookupBucket(impl_info);
   for (auto prev_impl_id : lookup_bucket_ref) {
     if (MergeImplRedecl(context, impl_info, prev_impl_id)) {
-      impl_decl.impl_id = prev_impl_id;
+      valid_redeclaration = IsValidImplRedecl(context, impl_info, prev_impl_id);
+      if (valid_redeclaration) {
+        impl_decl.impl_id = prev_impl_id;
+      }
       break;
     }
   }
 
   // Create a new impl if this isn't a valid redeclaration.
```

Comment:
FWIW I am finding it hard to follow the logic in this change. Specifically:
- There's now `valid_redeclaration` as a bool.
- The comment here shows we check if we set the `impl_decl.impl_id` to something valid to see if it's a valid redeclaration.
- The bool can be true without ever going into the point where we set `impl_decl.impl_id`. So the bool means something similar, but different.
- Then later we check the bool instead of `impl_decl.impl_id` to see if it's a "valid redeclaration".

Reply:
You are right, this was the result of iterating until it behaved the way I wanted. The purpose of this bool was to plumb the information about whether we've already issued a particular error to the point where we can avoid producing another error. I think it would make sense to rename this `!invalid_redeclaration`, but I don't know if that goes far enough. Does #4738 help?

Reply:
yes it helps a lot with the intent of the bool.

I do think there's still a bit of collision with "valid redeclaration" meaning that `impl_decl.impl_id.is_valid()`. Maybe `invalid_redeclaration` would be better using a different name than "valid" and IsValidImplRedecl() could be renamed to match. I am not sure but maybe "legal" or something for that, it's checking rules where "valid" is used to refer to the id value.